### PR TITLE
cranelift-interpreter: Implement a bunch of SIMD arithmetic ops

### PR DESCRIPTION
**Pull request by afonso360**
👋 Hey,

The interpreter can already deal with some opcodes in a polymorphic way, but this wasn't enabled for most binary and unary opcodes. This PR changes that. We get a bunch of opcodes implemented in a relatively easy way.

Please review this commit by commit, The first commits are just moving ops that weren't implemented from the main `simd-arithmetic.clif` into their own separate files. So that we could later enable that file for testing in the interpreter.

Will add fuzzer support in a separate PR.

### CODE REVIEW

File: cranelift/interpreter/src/step.rs [hunk 1]
```diff
@@ -199,45 +199,51 @@ where
...
+        |op: fn(V, V) -> ValueResult<V>, left: V, right: V| -> ValueResult<ControlFlow<V>> {
+            let ctrl_ty = inst_context.controlling_type().unwrap();
+            let res = binary_arith(left, right, ctrl_ty, op, false)?;
+            Ok(assign(res))
+        };
 
     // Same as `binary_unsigned`, but converts the values to their unsigned form before the
     // operation and back to signed form afterwards. Since Cranelift types have no notion of
     // signedness, this enables operations that depend on sign.
     let binary_unsigned =
```
Comment (jameysharp) [inline cranelift/interpreter/src/step.rs:222]: Pre-existing: should this comment have said "Same as `binary`"?

**Review by jameysharp — APPROVED** [review]
Very nice! I don't have any changes to suggest for this PR. Feel free to merge as-is.

I'm not totally happy with the existing signature of `binary_arith` though, or by extension the `unary_arith` function. Specifically, using a `bool` argument to indicate `unsigned` confused me during review; I couldn't figure out where the unsigned conversions had disappeared to in the `binary_unsigned` closure.

As future work, what do you think of this approach:
1. Extend `DataValue::convert` so that `ValueConversionKind::ToUnsigned` and `ValueConversionKind::ToSigned` return `self` unchanged if it already has the right type.
2. Pass a `ValueConversionKind` to `binary_arith`/`unary_arith` instead of the `unsigned` flag, and always call `convert` on each argument.
3. Optionally, get rid of the `binary_unsigned` and `binary_unsigned_can_trap` closures in favor of passing a `ValueConversionKind` to `binary` and `binary_can_trap`.

File: cranelift/interpreter/src/step.rs (hunk 1 above)
Comment (afonso360) [reply cranelift/interpreter/src/step.rs:222]: Yeah, I think so, I'll update this when doing the refactoring above ^

**Comment by afonso360** [comment]
@jameysharp It looks like some shift tests were added to `simd-arithmetic.clif` in the mean time, I've moved them into separate files because the interpreter is still broken on those operations.

Since this isn't a typo fix or anything like that would you be able to review again (these changes are in the latest commit)?

**Review by jameysharp — APPROVED** [review]
Sure, looks good to me!

**Comment by alexcrichton** [comment]
I don't know what happened there with the merge queue so I'm gonna put this back in and assume it was a fluke.

**Pull request merged**